### ISSUE TITLE
Training speed for transformer

### ISSUE THREAD

**Issue by vince62s** [marian-nmt/marian#145]
I looked at the benchmarks, obviously for RNNs.

Currently the speed for the default transformer on 1 GPU GTX 1080ti (close to Titan X, right ?)
is at 5500 words per second. My only change was -w 9000 to fit in the 11GB ram.
You confirm it's actually subwords, rights ?
But the speed looks low (both compared to T2T and to the fact it should be faster than your RNN baselines)
Do you have any numbers to compare ?

**Comment by emjotde** [comment]
Yes, source tokens. 
I haven't benchmarked T2T lately, it used to be a good deal faster than their implementation, might have changed. Will take a look soon. Do you have some numbers?

Also make sure to report w/s numbers the a later iteration. The first one will include things like memory allocation, collecting stats for batch-fitting etc. 

As for the speed improvement over the RNN baseline, I do not believe it is supposed to be faster than a shallow model, it is a good deal faster than a comparably deep RNN model. The Transformer model with depth 6 has about 12 layers in the encoder and 18 in the decoder, and multi-head attention is adding a bit of computation too.

**Comment by vince62s** [comment]
My last run on transformer_base (6 layers of 512 ...) on 4 GTX1080ti gives me 2.3 steps per second.
each step is 4 times the 4096 batch. That's 37683 token per second. Probably sublinear, but if I divide by 4 it's much higher than the current 5500 words per second.

**Comment by emjotde** [comment]
Dunno, maybe they improved something. Will check it out.

**Comment by vince62s** [comment]
no they didn't, it's been the same for months. I would just double check if what I am reporting for Marian transformer is close to your observations in term of numbers.

**Comment by emjotde** [comment]
It seems ok, I saw numbers around 22K wps on 4 GPUs. I just did not observe T2T to be that fast. I also believe the implementations are pretty similar, so that would be curious.

**Comment by vince62s** [comment]
oh wait ! by default I see dim-rnn=1024
is this also the layer soze of the transformer layers ?

**Comment by emjotde** [comment]
No, default is 512 for transformer. You are good with the settings from the example.

**Comment by vince62s** [comment]
might help to output to console/log the actual batch size of tokens.

**Comment by emjotde** [comment]
Give me some time to look into this. Cannot really comment without checking it out.

**Comment by vince62s** [comment]
Marcin,
I don't want to be a pain here. is there a gitter or slack for questions ? unless you prefer to get the questions here.
is it possible to pursue an interrupted training ? will it take automatically the last checkpoint ? recalculate the dataset index ?

**Comment by emjotde** [comment]
Hi,
for general questions you can use the google group, that's probably a better place than github issues. 

https://groups.google.com/forum/#!forum/marian-nmt

As for continuation. If you just repeat the command it will load the last checkpoint and continue from there. We do not have data set indices for resumption, I never saw the point of that.

**Comment by vince62s** [comment]
this is the reason why ....

[2017-12-21 15:17:07] Saving model to model/model.iter135000.npz
[2017-12-21 15:17:07] Saving model to model/model.npz
[2017-12-21 15:17:11] [valid] 135000 : cross-entropy : 56.8704 : new best
[2017-12-21 15:17:15] [valid] 135000 : perplexity : 7.7452 : new best
[2017-12-21 15:18:48] [valid] 135000 : translation : 22.21 : new best

interruption and restart

[2017-12-21 16:51:07] Saving model to model/model.iter140000.npz
[2017-12-21 16:51:12] Saving model to model/model.npz
[2017-12-21 16:51:19] [valid] 140000 : cross-entropy : 58.0938 : new best
[2017-12-21 16:51:23] [valid] 140000 : perplexity : 8.0939 : new best
[2017-12-21 16:53:18] [valid] 140000 : translation : 21.64 : new best

**Comment by emjotde** [comment]
Yeah, transformer model can ideed be a bit stuborn. We will take a look at that too. For RNN models it did usually no harm, there is actually a paper which shows that hard resets of Adam are helpful, apparently not for transformer.

**Comment by chrishokamp** [comment]
hi guys, sorry to hijack the thread, but Marcin's last comment implies that
Adam's params get reset when training is interrupted and then restarted
from a checkpoint. Do the Adam (or other optimizer) params get saved in
model checkpoints, or are they reset each time?

**Comment by emjotde** [comment]
Currently the Adam parameters are NOT being saved. We have plans to add this soon. I think there was some work in that direction recently, but not done yet. It's quite a pain to handle the different sharded optimizers for all multi-gpu settings.

**Comment by emjotde** [comment]
@snukky was looking into this, but I believe we all got hung up on NAACL now.

**Comment by chrishokamp** [comment]
Great, thanks for the clarification (and for the amazing project!)

**Comment by vince62s** [comment]
when looking at this:
[2017-12-22 08:57:27] Ep. 5 : Up. 264500 : Sen. 1443180 : Cost 106.73 : Time 220.67s : 5370.73 words/s : L.r. 7.3785e-05
[2017-12-22 09:01:08] Ep. 5 : Up. 265000 : Sen. 1483598 : Cost 104.36 : Time 221.71s : 5363.75 words/s : L.r. 7.3715e-05
If I do the math properly:
221 sec x 5370 = 1186770 words per 500 steps, ie 2373 token per batch
this is on 1 GPU for which I set -w 9000 to fit in the 11GB 1080ti RAM.

With T2T I can fit a batch_size of 4096 token in such a GPU.
Since the performance is very sensitive to batch_size is there an explanation for such a difference ?

thks

**Comment by vince62s** [comment]
Coming back to the original question, if I may ask:
When you bucket the batch by length, do you sort based on source or target length ? (or max of both)

**Comment by emjotde** [comment]
This has recently been fixed, it's now by both, first target then source. It actually made training 2x faster!

**Comment by vince62s** [comment]
héhé I knew it, great now I can really test the whole stuff.

**Comment by emjotde** [comment]
It was just target before, and source was random, stupid oversight. Still amazed though it made such a difference.

**Comment by emjotde** [comment]
BTW, resume should work much now too. There still small issues, but I believe they are hardly noticeable. I was able to continue interrupted transformer training runs without problems,

**Comment by vince62s** [comment]
yeah we saw a similar speed increase on ONMT-PY.
I'll test everything.
I think it should be possible to achieve the scores you show without all these tricks of ensembling R2L and L2R.
Just adding the back-translation to the main corpus was not enough ?
I was able to reach 28.4 on NT17 without BT.
I'll try anyway.
Great work.

**Comment by emjotde** [comment]
Quite possibly hyper-parameters or just the fact that I kill training after 7-8 epochs. I think the transformer still has the potential to improve quite a bit in the long tail. Also r2l does very little in the end. I can put my numbers for separate stages in there. For test2017, r2l helps only 0.2 BLEU over the ensemble which seems to be in-line with what you are saying.

**Comment by emjotde** [comment]
Numbers updated. All single systems slightly better than yours. Worth to train it a bit longer than 8 epochs for maxing out performance.

**Comment by vince62s** [comment]
I think the run-me.sh for wmt17-transformer is wrong with -type s2s instead of transformer

**Comment by emjotde** [comment]
That's only the small model for quick back-translation.

**Comment by emjotde** [comment]
I will add comments to the scripts later.

**Comment by vince62s** [comment]
hmm, are you sure about the x2 faster ? Am I missing something ? (on 4 GPU)
[2018-03-16 21:42:28] Ep. 1 : Up. 500 : Sen. 221065 : Cost 9.98 : Time 304.00s : 21383.61 words/s : L.r. 9.3750e-06
[2018-03-16 21:47:20] Ep. 1 : Up. 1000 : Sen. 440576 : Cost 8.54 : Time 291.85s : 22291.46 words/s : L.r. 1.8750e-05
[2018-03-16 21:52:11] Ep. 1 : Up. 1500 : Sen. 662548 : Cost 7.91 : Time 290.68s : 22328.39 words/s : L.r. 2.8125e-05

**Comment by emjotde** [comment]
Should be in there too, it's a pointer basically. Will take a look later. Can you post your config here?

**Comment by vince62s** [comment]
[2018-03-16 21:37:01] [config] after-batches: 0
[2018-03-16 21:37:01] [config] after-epochs: 0
[2018-03-16 21:37:01] [config] allow-unk: false
[2018-03-16 21:37:01] [config] batch-flexible-lr: false
[2018-03-16 21:37:01] [config] batch-normal-words: 1920
[2018-03-16 21:37:01] [config] beam-size: 6
[2018-03-16 21:37:01] [config] best-deep: false
[2018-03-16 21:37:01] [config] clip-norm: 5
[2018-03-16 21:37:01] [config] cost-type: ce-mean-words
[2018-03-16 21:37:01] [config] cpu-threads: 0
[2018-03-16 21:37:01] [config] data-weighting-type: sentence
[2018-03-16 21:37:01] [config] dec-cell: gru
[2018-03-16 21:37:01] [config] dec-cell-base-depth: 2
[2018-03-16 21:37:01] [config] dec-cell-high-depth: 1
[2018-03-16 21:37:01] [config] dec-depth: 6
[2018-03-16 21:37:01] [config] devices:
[2018-03-16 21:37:01] [config]   - 0
[2018-03-16 21:37:01] [config]   - 1
[2018-03-16 21:37:01] [config]   - 2
[2018-03-16 21:37:01] [config]   - 3
[2018-03-16 21:37:01] [config] dim-emb: 512
[2018-03-16 21:37:01] [config] dim-rnn: 1024
[2018-03-16 21:37:01] [config] dim-vocabs:
[2018-03-16 21:37:01] [config]   - 0
[2018-03-16 21:37:01] [config]   - 0
[2018-03-16 21:37:01] [config] disp-freq: 500
[2018-03-16 21:37:01] [config] dropout-rnn: 0
[2018-03-16 21:37:01] [config] dropout-src: 0
[2018-03-16 21:37:01] [config] dropout-trg: 0
[2018-03-16 21:37:01] [config] early-stopping: 10
[2018-03-16 21:37:01] [config] embedding-fix-src: false
[2018-03-16 21:37:01] [config] embedding-fix-trg: false
[2018-03-16 21:37:01] [config] embedding-normalization: false
[2018-03-16 21:37:01] [config] enc-cell: gru
[2018-03-16 21:37:01] [config] enc-cell-depth: 1
[2018-03-16 21:37:01] [config] enc-depth: 6
[2018-03-16 21:37:01] [config] enc-type: bidirectional
[2018-03-16 21:37:01] [config] exponential-smoothing: 0.0001
[2018-03-16 21:37:01] [config] grad-dropping-momentum: 0
[2018-03-16 21:37:01] [config] grad-dropping-rate: 0
[2018-03-16 21:37:01] [config] grad-dropping-warmup: 100
[2018-03-16 21:37:01] [config] guided-alignment-cost: ce
[2018-03-16 21:37:01] [config] guided-alignment-weight: 1
[2018-03-16 21:37:01] [config] ignore-model-config: false
[2018-03-16 21:37:01] [config] keep-best: false
[2018-03-16 21:37:01] [config] label-smoothing: 0.1
[2018-03-16 21:37:01] [config] layer-normalization: false
[2018-03-16 21:37:01] [config] learn-rate: 0.0003
[2018-03-16 21:37:01] [config] log: model/train.log
[2018-03-16 21:37:01] [config] log-level: info
[2018-03-16 21:37:01] [config] lr-decay: 0
[2018-03-16 21:37:01] [config] lr-decay-freq: 50000
[2018-03-16 21:37:01] [config] lr-decay-inv-sqrt: 16000
[2018-03-16 21:37:01] [config] lr-decay-repeat-warmup: false
[2018-03-16 21:37:01] [config] lr-decay-reset-optimizer: false
[2018-03-16 21:37:01] [config] lr-decay-start:
[2018-03-16 21:37:01] [config]   - 10
[2018-03-16 21:37:01] [config]   - 1
[2018-03-16 21:37:01] [config] lr-decay-strategy: epoch+stalled
[2018-03-16 21:37:01] [config] lr-report: true
[2018-03-16 21:37:01] [config] lr-warmup: 16000
[2018-03-16 21:37:01] [config] lr-warmup-at-reload: false
[2018-03-16 21:37:01] [config] lr-warmup-cycle: false
[2018-03-16 21:37:01] [config] lr-warmup-start-rate: 0
[2018-03-16 21:37:01] [config] max-length: 100
[2018-03-16 21:37:01] [config] max-length-crop: false
[2018-03-16 21:37:01] [config] maxi-batch: 1000
[2018-03-16 21:37:01] [config] maxi-batch-sort: trg
[2018-03-16 21:37:01] [config] mini-batch: 64
[2018-03-16 21:37:01] [config] mini-batch-fit: true
[2018-03-16 21:37:01] [config] mini-batch-fit-step: 10
[2018-03-16 21:37:01] [config] mini-batch-words: 0
[2018-03-16 21:37:01] [config] model: model/model.npz
[2018-03-16 21:37:01] [config] multi-node: false
[2018-03-16 21:37:01] [config] multi-node-overlap: true
[2018-03-16 21:37:01] [config] n-best: false
[2018-03-16 21:37:01] [config] no-reload: false
[2018-03-16 21:37:01] [config] no-restore-corpus: false
[2018-03-16 21:37:01] [config] no-shuffle: false
[2018-03-16 21:37:01] [config] normalize: 0.6
[2018-03-16 21:37:01] [config] optimizer: adam
[2018-03-16 21:37:01] [config] optimizer-delay: 1
[2018-03-16 21:37:01] [config] optimizer-params:
[2018-03-16 21:37:01] [config]   - 0.9
[2018-03-16 21:37:01] [config]   - 0.98
[2018-03-16 21:37:01] [config]   - 1e-09
[2018-03-16 21:37:01] [config] overwrite: false
[2018-03-16 21:37:01] [config] quiet: false
[2018-03-16 21:37:01] [config] quiet-translation: true
[2018-03-16 21:37:01] [config] relative-paths: false
[2018-03-16 21:37:01] [config] right-left: false
[2018-03-16 21:37:01] [config] save-freq: 5000
[2018-03-16 21:37:01] [config] seed: 1111
[2018-03-16 21:37:01] [config] skip: false
[2018-03-16 21:37:01] [config] sqlite: ""
[2018-03-16 21:37:01] [config] sqlite-drop: false
[2018-03-16 21:37:01] [config] sync-sgd: true
[2018-03-16 21:37:01] [config] tempdir: /tmp
[2018-03-16 21:37:01] [config] tied-embeddings: false
[2018-03-16 21:37:01] [config] tied-embeddings-all: true
[2018-03-16 21:37:01] [config] tied-embeddings-src: false
[2018-03-16 21:37:01] [config] train-sets:
[2018-03-16 21:37:01] [config]   - data/corpus.bpe.en
[2018-03-16 21:37:01] [config]   - data/corpus.bpe.de
[2018-03-16 21:37:01] [config] transformer-dim-ffn: 2048
[2018-03-16 21:37:01] [config] transformer-dropout: 0.1
[2018-03-16 21:37:01] [config] transformer-dropout-attention: 0
[2018-03-16 21:37:01] [config] transformer-dropout-ffn: 0
[2018-03-16 21:37:01] [config] transformer-ffn-activation: swish
[2018-03-16 21:37:01] [config] transformer-heads: 8
[2018-03-16 21:37:01] [config] transformer-postprocess: dan
[2018-03-16 21:37:01] [config] transformer-postprocess-emb: d
[2018-03-16 21:37:01] [config] transformer-preprocess: ""
[2018-03-16 21:37:01] [config] type: transformer
[2018-03-16 21:37:01] [config] valid-freq: 5000
[2018-03-16 21:37:01] [config] valid-log: model/valid.log
[2018-03-16 21:37:01] [config] valid-max-length: 1000
[2018-03-16 21:37:01] [config] valid-metrics:
[2018-03-16 21:37:01] [config]   - ce-mean-words
[2018-03-16 21:37:01] [config]   - perplexity
[2018-03-16 21:37:01] [config]   - translation
[2018-03-16 21:37:01] [config] valid-mini-batch: 64
[2018-03-16 21:37:01] [config] valid-script-path: ./scripts/validate.sh
[2018-03-16 21:37:01] [config] valid-sets:
[2018-03-16 21:37:01] [config]   - data/valid.bpe.en
[2018-03-16 21:37:01] [config]   - data/valid.bpe.de
[2018-03-16 21:37:01] [config] valid-translation-output: data/valid.bpe.en.output
[2018-03-16 21:37:01] [config] vocabs:
[2018-03-16 21:37:01] [config]   - model/vocab.ende.yml
[2018-03-16 21:37:01] [config]   - model/vocab.ende.yml
[2018-03-16 21:37:01] [config] workspace: 9000

**Comment by emjotde** [comment]
Set `--mini-batch 1000` and `--maxi-batch 1000`. That will preload 1M sentences which can then be properly packed. Currently it is only preloading 64K. Since you use `--mini-batch-fit` is will ignore the mini-batch size for actual batching and only use it for preloading. 

This is something I need to solve in a better way, probably by switching `--maxi-batch` to an absolute sentences-based value instead of a multiplier of mini-batch size.

**Comment by vince62s** [comment]
by doing so I have to drastically reduce the workspace. will that reduce a lot the mini-batch size ?

**Comment by emjotde** [comment]
Mini-batch-fit should take care of all. Cannot see how that would affect workspace at all.

**Comment by emjotde** [comment]
Yes. That is an awful hack currently.

**Comment by vince62s** [comment]
ok, a bit better, not twice though.
[2018-03-17 15:28:54] [memory] Reserving 119 MB, device gpu1
[2018-03-17 15:34:28] Ep. 1 : Up. 500 : Sen. 321439 : Cost 9.95 : Time 392.11s : 23987.32 words/s : L.r. 9.3750e-06
[2018-03-17 15:40:14] Ep. 1 : Up. 1000 : Sen. 638951 : Cost 8.52 : Time 345.61s : 27902.33 words/s : L.r. 1.8750e-05
[2018-03-17 15:45:57] Ep. 1 : Up. 1500 : Sen. 967284 : Cost 7.93 : Time 343.14s : 27675.67 words/s : L.r. 2.8125e-05

**Comment by emjotde** [comment]
You can also add `--mini-batch-fit-step 2` which will pack the mini-batches with a smaller granularity. Default is 10.

**Comment by vince62s** [comment]
another bit
[2018-03-17 17:33:29] Ep. 1 : Up. 2500 : Sen. 1844265 : Cost 7.14 : Time 381.98s : 28215.39 words/s : L.r. 4.6875e-05
[2018-03-17 17:39:53] Ep. 1 : Up. 3000 : Sen. 2203856 : Cost 6.83 : Time 384.67s : 28042.07 words/s : L.r. 5.6250e-05
[2018-03-17 17:46:16] Ep. 1 : Up. 3500 : Sen. 2566015 : Cost 6.60 : Time 382.86s : 28550.86 words/s : L.r. 6.5625e-05
